src/middlewared/middlewared/plugins/catalogs_linux/items.py

```diff
@@ -130,52 +130,68 @@ def items(self, label, options):
-        # This allows us to use these folders for placing helm library charts and docs respectively
-        trains = {'charts': {}, 'test': {}}
-        options = options or {}
-        all_trains = options.get('all_trains', True)
-        trains_filter = options.get('trains', [])
-        questions_context = self.middleware.call_sync('catalog.get_normalised_questions_context')
-        unhealthy_apps = set()
-        if options.get('alert') and options.get('label'):
-            preferred_trains = self.middleware.call_sync('catalog.get_instance', options['label'])['preferred_trains']
-        else:
-            preferred_trains = []
-
+    def retrieve_train_names(self, location, all_trains=True, trains_filter=None):
+        train_names = []
+        trains_filter = trains_filter or []
         for train in os.listdir(location):
             if (
                 not (all_trains or train in trains_filter) or not os.path.isdir(
                     os.path.join(location, train)
                 ) or train.startswith('.') or train in ('library', 'docs') or not VALID_TRAIN_REGEX.match(train)
             ):
                 continue
+            train_names.append(train)
+        return train_names
 
+    @private
+    def get_trains(self, job, location, catalog, options):
+        # We make sure we do not dive into library and docs folders and not consider those a train
+        # This allows us to use these folders for placing helm library charts and docs respectively
+        trains = {'charts': {}, 'test': {}}
+        questions_context = self.middleware.call_sync('catalog.get_normalised_questions_context')
+        unhealthy_apps = set()
+        preferred_trains = catalog['preferred_trains']
+
+        trains_to_traverse = self.retrieve_train_names(location, options['retrieve_all_trains'], options['trains'])
+        # In order to calculate job progress, we need to know number of items we would be traversing
+        items = {}
+        for train in trains_to_traverse:
+            items.update({
+                i: train for i in os.listdir(os.path.join(location, train))
+                if os.path.isdir(os.path.join(location, train, i))
+            })
+
+        job.set_progress(8, f'Retrieving {", ".join(trains_to_traverse)!r} train(s) information')
+
+        total_items = len(items)
+        for index, item in enumerate(items):
+            train = items[item]
             trains[train] = {}
-            category_path = os.path.join(location, train)
-            for item in filter(lambda p: os.path.isdir(os.path.join(category_path, p)), os.listdir(category_path)):
-                item_location = os.path.join(category_path, item)
-                if not os.path.isdir(item_location):
-                    continue
+            item_location = os.path.join(location, train, item)
+            job.set_progress(
+                ((index / total_items) * (90 - 10)) + 10,
```

Comment:
1. Why `(90 - 10)`? Just make it `* 80`
2. You probably want to cast this to an `int()` or `round()` it to 2 decimal places so you don't have a `33.3333334%` or something like that.